src/rgw/services/svc_sys_obj_core.h

```diff
@@ -0,0 +1,208 @@
+    has_attrs = rhs.has_attrs;
+    exists = rhs.exists;
+    size = rhs.size;
+    mtime = rhs.mtime;
+    epoch = rhs.epoch;
+    if (rhs.obj_tag.length()) {
+      obj_tag = rhs.obj_tag;
+    }
+    has_data = rhs.has_data;
+    if (rhs.data.length()) {
+      data = rhs.data;
+    }
+    prefetch_data = rhs.prefetch_data;
+    pg_ver = rhs.pg_ver;
+    objv_tracker = rhs.objv_tracker;
+  }
+};
+
+class RGWSysObjectCtxBase {
+  std::map<rgw_raw_obj, RGWSysObjState> objs_state;
+  RWLock lock;
+
+public:
+  explicit RGWSysObjectCtxBase() : lock("RGWSysObjectCtxBase") {}
+
+  RGWSysObjectCtxBase(const RGWSysObjectCtxBase& rhs) : objs_state(rhs.objs_state),
+                                                  lock("RGWSysObjectCtxBase") {}
+  RGWSysObjectCtxBase(const RGWSysObjectCtxBase&& rhs) : objs_state(std::move(rhs.objs_state)),
+                                                   lock("RGWSysObjectCtxBase") {}
+
+  RGWSysObjState *get_state(const rgw_raw_obj& obj) {
+    RGWSysObjState *result;
+    std::map<rgw_raw_obj, RGWSysObjState>::iterator iter;
+    lock.get_read();
+    assert (!obj.empty());
+    iter = objs_state.find(obj);
+    if (iter != objs_state.end()) {
+      result = &iter->second;
+      lock.unlock();
+    } else {
+      lock.unlock();
+      lock.get_write();
+      result = &objs_state[obj];
+      lock.unlock();
+    }
+    return result;
+  }
+
+  void set_atomic(rgw_raw_obj& obj) {
+    RWLock::WLocker wl(lock);
+    assert (!obj.empty());
+  }
+  void set_prefetch_data(rgw_raw_obj& obj) {
+    RWLock::WLocker wl(lock);
+    assert (!obj.empty());
+    objs_state[obj].prefetch_data = true;
+  }
+  void invalidate(rgw_raw_obj& obj) {
+    RWLock::WLocker wl(lock);
+    auto iter = objs_state.find(obj);
+    if (iter == objs_state.end()) {
+      return;
+    }
+    objs_state.erase(iter);
+  }
+};
+
+class RGWS_SysObj_Core : public RGWService
+{
+public:
+  RGWS_SysObj_Core(CephContext *cct) : RGWService(cct, "sysobj_core") {}
+
+  int create_instance(const std::string& conf, RGWServiceInstanceRef *instance) override;
+};
+
+class RGWSI_SysObj_Core : public RGWServiceInstance
+{
+  friend class RGWSI_SysObj;
+
+protected:
+  std::shared_ptr<RGWSI_RADOS> rados_svc;
+  std::shared_ptr<RGWSI_Zone> zone_svc;
+
+  struct GetObjState {
+    RGWSI_RADOS::Obj rados_obj;
+    bool has_rados_obj{false};
+    uint64_t last_ver{0};
+
+    GetObjState() {}
+
+    int get_rados_obj(RGWSI_RADOS *rados_svc,
+                      RGWSI_Zone *zone_svc,
+                      rgw_raw_obj& obj,
+                      RGWSI_RADOS::Obj **pobj);
+  };
+
+
+  std::map<std::string, RGWServiceInstance::dependency> get_deps() override;
+  int load(const std::string& conf, std::map<std::string, RGWServiceInstanceRef>& dep_refs) override;
+
+  int get_rados_obj(RGWSI_Zone *zone_svc, rgw_raw_obj& obj, RGWSI_RADOS::Obj *pobj);
+
+  virtual int raw_stat(rgw_raw_obj& obj, uint64_t *psize, real_time *pmtime, uint64_t *epoch,
```

Comment:
it looks like obj could be a const ref here? same with the other functions here

Reply:
well, obj could be canonicalized somehwere below, but it's time to change this. Will reject objects with empty oids (artifact of a very old scheme).